### DISCUSSION TITLE
enhanced:img not working at all for me

### DISCUSSION THREAD

**Discussion by StokleyO** [sveltejs/kit#11098] [Q&A]
Hey everyone,

I've recently begun attempting to use Svelte's new enhanced:img feature as described here (https://kit.svelte.dev/docs/images#sveltejs-enhanced-img), however, I am not getting the desired results. After running the function to install @sveltejs/enhanced-img and adding enhancedImages to my vite.config.js file, anytime I try to use the <enhanced:img /> tag, I get an error saying 'Cannot read properties of undefined (reading 'src')'. Then also, when I try to add ?enhanced to the end of my dynamic photo import, I get an error that says no such file exists and then it shows me the exact path to get to the file. Anyone been able to implement this successfully yet? The specific image here is a png file. And I'm using SvelteKit 1.20.4 and Svelte 4

**Comment by benmccann** [comment]
@StokleyO there's not enough information here to tell where things might be going wrong. Can you share a repository that reproduces the error?

**Reply by StokleyO** [reply]
@benmccann Thank you for responding. I can't share my repo since it's a private repo, but I can show some screenshots. It's relatively simple what I'm trying to accomplish. I have a .png file that I'd like to pass through the enhanced img component. I followed the documentation, and from what I see, there's no reason it should be failing
![Hero enhanced img](https://github.com/sveltejs/kit/assets/83097972/8a26ffa2-82b4-4522-82d2-19c66323497d)
![vite config](https://github.com/sveltejs/kit/assets/83097972/09b12fca-05de-47be-baea-c8f8ad3c5118)
![error I receive](https://github.com/sveltejs/kit/assets/83097972/d8b1eadc-c5e3-4e0a-8a8e-5584154aad12)

The file that it's claiming doesn't exist, absolutely does exist, and when I replace enhanced:img with a regular img tag, and then remove ?enhanced from the path, the image loads fine. Do I have to run npm build before npm run dev?

**Reply by StokleyO** [reply]
If it helps, this is the error I receive in my console
![Console error](https://github.com/sveltejs/kit/assets/83097972/50360fad-8785-4c46-b3ee-96b1ac2c71f2)

**Reply by benmccann** [reply]
If you cannot share your original code, then please recreate the error in a new repository and share that

**Reply by StokleyO** [reply]
@benmccann While I set that up, is it possibly a dependency issue? Maybe a version of Vite that's incompatible?
![dependencies](https://github.com/sveltejs/kit/assets/83097972/44241e47-3a61-469d-a880-fdc3f0cdda9e)

**Reply by benmccann** [reply]
No

**Reply by StokleyO** [reply]
@benmccann Here is my repo (https://github.com/StokleyO/vaulted) I just checked out the video you recently did about enhanced:img and for the life of me I don't see what I'm doing wrong. My images are in a src/lib/assets folder. Is this possibly the reason?

**Edit: The post below me would be a better representation of the issue**

**Reply by thedivtagguy** [reply]
I have the same issue. I would really like to be able to use this but I can't get it working. I've followed the docs. Sharing my repo and the page where I'm using it [here](https://github.com/thedivtagguy/site/blob/master/src/lib/components/Homepage/Eymen.svelte). I get this error too: 
![image](https://github.com/sveltejs/kit/assets/6006888/36324b5a-38e2-4932-a39a-049d2244ed3b)

The file does exist and if I import it without `?enhanced` and within a normal `<img>` tag, it does work.

**Reply by saschakrueger** [reply]
I am having exactly the same problem, but only on my desktop machine. When I work on my laptop it just works fine with the same repo, so I figure it must have something to do with the environment. Did not have the time to check in detail yet, but at least I can tell that I am using the exact same node version on both machines.
Only noticeable difference so far is the Mac OS version. (13.6.1 not working vs. 14.0 working), but I don't think that this is the issue.

**Reply by thedivtagguy** [reply]
Well fwiw, if it helps, I'm on Linux (Pop OS!).
```
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Pop!_OS 22.04 LTS

node --version
v20.4.0
```

**Reply by benmccann** [reply]
@StokleyO I get a 404 on your repo https://github.com/StokleyO/vaulted

**Reply by benmccann** [reply]
@thedivtagguy here's how I set it up on your site. It worked for me. Had you set it up differently?

![Screenshot from 2023-11-29 19-54-35](https://github.com/sveltejs/kit/assets/322311/a685dd4f-d799-4c1c-9af7-154098e28d69)

**Reply by StokleyO** [reply]
@benmccann I'm sorry about that, you should be able to access it now. I figured it would be easier for you to look at the above commenter since he linked his repo more cleanly

**Reply by thedivtagguy** [reply]
@benmccann Apologies, I should have permalinked to that commit. I did in fact use it the same way but it was causing a 500 error and preventing builds so I had to remove it; when you saw it the component must have been updated. I've added it back in and [here's how I am using it](https://github.com/thedivtagguy/site/blob/53a30cdcb0940ed9fe1abe71e7122da6f8c8ca38/src/lib/components/Homepage/Eymen.svelte). 

I think it is the same and even my vite.config is in order. Is there any other system information I should share? 

```
[Error: ENOENT: no such file or directory, stat '/home/amnbh/Desktop/Dev%20Projects/website/src/lib/assets/images/common/eymen-1.png'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/home/amnbh/website/src/lib/assets/images/common/eymen-1.png'
}
```

**Reply by StokleyO** [reply]
@benmccann I've been trying to implement it for [the hero on my home page](https://github.com/StokleyO/vaulted/blob/main/src/lib/components/HeroSection.svelte)

**Reply by benmccann** [reply]
@thedivtagguy does it happen for you when you run `pnpm dev` or `pnpm build`? I pulled your changes and `pnpm dev` works fine for me. I can't run `pnpm build` because it needs credentials for a CMS

**Reply by thedivtagguy** [reply]
@benmccann It happens even when I'm developing on localhost. Once I had the `enhanced` tags and hit save, the `500` error shows up and I can't continue even before building (so I guess even on `pnpm run dev` to answer your question).

**Reply by benmccann** [reply]
Hmm. This is very weird. @saschakrueger do you have the same version of Node.js installed on both machines? And does your project have a lockfile? (e.g. `package-lock.json`, `pnpm-lock.yaml`, etc.)

**Reply by saschakrueger** [reply]
@benmccann yes, both machines are running node 20.10.0 and yes, the project has a `package-lock.json`
I also tried to delete the lock-file and the `node_modules` folder and cleared the cache, but this did not help.

Also: `npm run build` works on both machines.
Only `npm run dev` throws the error on the desktop mac.

**Reply by thedivtagguy** [reply]
@benmccann Oh this is weird, I notice that the builds are passing on Netlify so I do see the transformed `avif` files and the generated `srcset`, so the only place it is failing is locally. I tried again because of @saschakrueger's reply so the earlier build failure must have been something else.

**Reply by thedivtagguy** [reply]
Anything I can do in the meantime to prevent this from happening while developing?

**Reply by saschakrueger** [reply]
I spend hours to find a workaround for this, but now I ended up using just a simple `img`-tag when in dev mode, just to check if dimensions and positioning of the images are correct, but this is obviously far from an acceptable solution. Still hoping that anyone comes up with a smart idea or even a fix for the problem.

**Reply by thedivtagguy** [reply]
I'm doing the same right now. But as the site grows, it'll become difficult to keep track of all the places I have to make the change to :/

**Reply by benmccann** [reply]
I wonder if there might be some race condition that's triggered more frequently on some machines than others. Does it ever alternate between succeeding and failing on the same machine with the same command?

**Reply by saschakrueger** [reply]
Well, at least in my case it never worked on the one machine and it always works on the other machine.

**Reply by StokleyO** [reply]
@benmccann I haven't been able to get it to work on my machine yet. When I run 'npm run dev' the images are broken. I haven't tried pushing to prod to see if Netlify can build it

**Reply by saschakrueger** [reply]
@benmccann I think the full error message has never been shared. So I attached a screen grab. Had to cover some of the path names because of confidential customer data, but you can see the full message anyway.
![Screenshot 2023-12-08 at 08 19 50](https://github.com/sveltejs/kit/assets/90637863/f2198eb4-3350-43ac-bd16-6f8d0d7ff591)

**Reply by StokleyO** [reply]
@benmccann Any thoughts on what may be causing the issue?

**Reply by webel** [reply]
I had this issue when trying to string interpolate the src and now with a dynamic import, it does work with static images.

**Reply by StokleyO** [reply]
Even when I use static image imports I still can not get this feature to work

**Reply by benmccann** [reply]
Thank you @saschakrueger. That was a big help.

Can everyone here update to `@sveltejs/enhanced-img` version `^0.1.7` and let me know if it helped or not?

**Reply by StokleyO** [reply]
@benmccann hey Ben, thanks for the reply! I've updated to 0.1.7. Image is now showing up for me, but it completely disfigures the proportions. Any feedback on this?
![image](https://github.com/sveltejs/kit/assets/83097972/62d179a2-fee8-4458-bc01-7c0814db69be)

**Reply by StokleyO** [reply]
@benmccann this is what it looks like without enhanced:img
![image](https://github.com/sveltejs/kit/assets/83097972/7528bd29-611f-4166-937a-0020042abf37)

**Reply by saschakrueger** [reply]
@benmccann thanks for looking into this. Updating to ^0.1.7 indeed fixed it kind of for me.


```
  <enhanced:img 
   src='/src/lib/assets/images/hero.jpg'
   sizes="(min-width:1920px) 1280px, (min-width:1080px) 640px, (min-width:768px) 400px"
   alt={alt} 
  />
```
works, but if I use something like `src={'/src/lib/assets/images/' + scr}` where `src` is a prop of my component, I receive a server error 500.


 ` <img src={'/src/lib/assets/images/' + scr} alt={alt} />` does work though.

@StokleyO : This looks rather like an issue with your css. You probably used an `img`-tag for your temporary fix, but now `enhanced:img` will generate a `picture`-tag for you.
Try to set `height: auto;` on the `img`-tag.

**Reply by StokleyO** [reply]
@saschakrueger you are correct, setting height: auto; resolves the issue. One last question, the image looks a bit too low quality on large & xl screens. can I set a custom width for those screen sizes in css?

**Reply by saschakrueger** [reply]
@StokleyO this is exactly what `sizes="(min-width:1920px) 1280px, (min-width:1080px) 640px, (min-width:768px) 400px"` is for. Within a `picture` tag it will use different size images for different screen sizes.  Above the largest image would be 1280px wide at a screen size of minimum 1920px. You could easily increase the values as needed, but keep in mind that that will also increase the load time of your app. Using modern formats like webp and avif will help to keep files small, even if you have rather large images.

**Reply by StokleyO** [reply]
@saschakrueger perfect!! thank you so much Sascha, you were a huge help. Thank you @benmccann as well

**Comment by victoriusnet** [comment]
@saschakrueger have you managed to solve the prop src issue?

I'm struggling with the same, reported here:
https://github.com/sveltejs/kit/issues/11535

Thank you!

**Reply by benmccann** [reply]
You were using the wrong syntax. I replied in the issue

**Comment by ghost** [comment]
After 10+ different errors I'll just remove this and stick to optimizing assets myself. This does not seem like a stable, production-ready tool right now.

**Comment by xdeus** [comment]
It seems **dynamically importing** images for <enhanced:img /> tag has **very limited functionality**. If you import an image, consider to use **if condition** before using the image in the html as below:

```
<script> import image from 'path-to-image/image.jpg?enhanced'; </script>
// ?enhanced query is must
```

```
{# if image !== undefined }
<enhanced:img src={image} />
{/if}
```

This worked for me. But it is **best to use static images** directly on <enhanced:img> tag, because it allows setting the dimensions of the image and you get the actual resized images. Using css to adjust width and height doesn't seemed to actually resize the image.

It happened in my case, you may directly check yourself. Do let us know what have you discovered.

**Comment by teemingc** [comment]
> As a build plugin, @sveltejs/enhanced-img can only optimize files located on your machine during the build process. If you have an image located elsewhere (such as a path served from your database, CMS, or backend), please read about [loading images dynamically from a CDN](https://kit.svelte.dev/docs/images#loading-images-dynamically-from-a-cdn).

https://github.com/sveltejs/kit/issues/11535#issuecomment-2026346534
